ros/angel_system_nodes/angel_system_nodes/hl2ss_ros_bridge.py

```diff
@@ -0,0 +1,316 @@
+        """
+        # Stop frame publishing thread
+        self._pv_active.clear()  # make RT active flag "False"
+        self._pv_thread.join()
+        self.get_logger().info("PV thread closed")
+
+        # Stop SI publishing thread
+        self._si_active.clear()  # make RT active flag "False"
+        self._si_thread.join()
+        self.get_logger().info("SI thread closed")
+
+        # Close client connections
+        self.hl2ss_pv_client.close()
+        hl2ss.stop_subsystem_pv(self.ip_addr, self.pv_port)
+        self.get_logger().info("PV client disconnected")
+
+        self.hl2ss_si_client.close()
+        self.get_logger().info("SI client disconnected")
+
+
+    def pv_publisher(self) -> None:
+        """
+        Main thread that gets frames from the HL2SS PV client and publishes
+        them to the image topic.
+        """
+        while self._pv_active.wait(0):  # will quickly return false if cleared.
+            data = self.hl2ss_pv_client.get_next_packet()
+
+            try:
+                image_msg = BRIDGE.cv2_to_imgmsg(data.payload, encoding="bgr8")
+                image_msg.header.stamp = hl2ss_stamp_to_ros_time(data.timestamp)
+                image_msg.header.frame_id = "PVFramesRGB"
+            except TypeError as e:
+                self.get_logger().warning(f"{e}")
+
+            self.frame_publisher.publish(image_msg)
+
+
+    def si_publisher(self) -> None:
+        """
+        Thread the gets spatial input packets from the HL2SS SI client, converts
+        the SI data to ROS messages, and publishes them.
+
+        Currently only publishes hand tracking data. However, eye gaze data and
+        head pose data is also available in the SI data packet.
+        """
+        while self._si_active.wait(0):  # will quickly return false if cleared.
+            data = self.hl2ss_si_client.get_next_packet()
+            si_data = hl2ss.unpack_si(data.payload)
+
+            # Publish the hand tracking data if it is valid
+            if si_data.is_valid_hand_left():
+                hand_msg_left = self.create_hand_pose_msg_from_si_data(
+                    si_data, "Left", data.timestamp
+                )
+                self.hand_publisher.publish(hand_msg_left)
+            if si_data.is_valid_hand_right():
+                hand_msg_right = self.create_hand_pose_msg_from_si_data(
+                    si_data, "Right", data.timestamp
+                )
+                self.hand_publisher.publish(hand_msg_right)
+
+
+    def create_hand_pose_msg_from_si_data(
+        self,
+        si_data: hl2ss.unpack_si,
+        hand: str,
+        timestamp: int
+    ) -> HandJointPosesUpdate:
+        """
+        Extracts the hand joint poses data from the HL2SS SI structure
+        and forms a ROS HandJointPosesUpdate message.
+        """
+        # List containing joint names that matches the ordering in the HL2SS
+        # SI_HandJointKind class
+        joint_list = [
```

Comment:
Part of me thinks this should be segregated out into a configuration file that this node is pointed to. Something lower bar would be to make this list a global variable instead of something that is created every time this method is called.